### PR TITLE
Implementation of speculative devirtualization for class protocols

### PR DESCRIPTION
**Pull request by swiftix**
#### What's in this pull request?

Implementation of speculative devirtualization for class protocols.

THIS IS WORK IN PROGRESS! DO NOT MERGE YET!
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [x] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by swiftix** [comment]
@swift-ci Please test

File: lib/IRGen/GenMeta.cpp [hunk 1]
```diff
@@ -4431,7 +4431,11 @@ llvm::Value *irgen::emitHeapMetadataRefForHeapObject(IRGenFunction &IGF,
     return emitLoadOfHeapMetadataRef(IGF, object,
                                      getIsaEncodingForType(IGF.IGM, objectType),
                                      suppressCast);
-
+  // Is it an opened class existential?
+  if (objectType.isAnyClassReferenceType() && isa<ArchetypeType>(objectType))
```
Comment (swiftix) [inline lib/IRGen/GenMeta.cpp]: @rjmccall John, is this a proper way to check that emitLoadOfHeapMetadataRef could be used for class existentials? Or do I need to add something else, e.g. to check if it is not a tagged pointer or something like this?

Comment (rjmccall) [inline lib/IRGen/GenMeta.cpp]: I actually completely forgot about tagged pointers.  Yes, you will need to check for them dynamically.

A more efficient and idiomatic way to do this check would be to dyn_cast to ArchetypeType and call requiresClass().

Note that the code you've written isn't specific to opened class existentials — it'll actually trigger on *any* class-bounded archetype.  That's probably okay, but the comment is wrong.

**Comment by rudkx** [comment]
@swiftix Is this still being worked on or can we close it out?

**Comment by swiftix** [comment]
We can close it for now. If we ever need it again, we can revive it.

**Pull request closed without merging**